### DISCUSSION TITLE
Discuss versioning

### DISCUSSION THREAD

**Discussion by krzema12** [krzema12/snakeyaml-engine-kmp#318] [General]
note: Since this change introduces a new enum entry we, probably, should update the minor version of the library (this change might break the code that uses `when` statement with all enum entries)

_Originally posted by @OptimumCode in https://github.com/krzema12/snakeyaml-engine-kmp/pull/313#discussion_r1899510482_

**Comment by krzema12** [comment]
@aSemy @OptimumCode I'm about to release a new version of the lib. The question about versioning is back.

snakeyaml-engine doesn't follow SemVer - I recall some breaking changes that weren't communicated as major version bump. SemVer is generally not perfect, or rather a definition of a breaking change is hard to formalize. In theory, every change can be a breaking one for some client. That's why usually for a given project, one has to define what constitutes a public API. For snakeyaml-engine, I noticed that some security-related patches that are breaking weren't communicated with a major version bump.

The question is what we want to do here. For the upcoming release, we can see this change: https://github.com/krzema12/snakeyaml-engine-kmp/commit/908f98434610a4fa54dffde73cecb9c09f288f65 which includes adding a new enum item to `ScalarStyle`, and if someone does an exhaustive `when` over this enum, their code will stop compiling. That's why I claim we should bump the version to 4.0.0.

There are some other nuances here, but I'll stop here to learn your thoughts!

**Reply by OptimumCode** [reply]
Some time ago I watched a recording of a talk from Joker conference related to versioning. And there was one phrase that, I think, really applicable here:

> If a patch version is incremented - compatibility is not broken (hopefully), if minor version is incremented - compatibility is broken a little, if major version is incremented - compatibility is broken a lot.

So, I would vote for incrementing the minor version and making a 3.1.0 release. The new enum entry doesn't sound to me like a good reason to create a new major version.

However, I completely understand why you would want to create a new major release (since it might break the compatibility). And I have no objections to that except the one from previous paragraphs.

**Reply by krzema12** [reply]
That's an interesting take!

The best way to check for breaking changes is to check the consumers. Of course it's impossible to analyze all consumers, but the most important one is probably kaml. I checked if kaml depends on exhaustively using the values of this enum, and probably not. I found this enum being used only here: https://github.com/charleskorn/kaml/blob/1b7766ec4195156027e01656bd151fd86048052d/src/commonMain/kotlin/com/charleskorn/kaml/YamlOutput.kt#L225-L240

So kaml will probably be fine if we release 3.1.0.

Before releasing, let me work a few more days on https://github.com/krzema12/snakeyaml-engine-kmp/pull/314 to see if it can be included in the release, this one is tricky. If you have some bandwidth, I'd appreciate some help - see my notes in the PR.
